internal/backup/volumesnapshot_action.go

```diff
@@ -149,13 +153,78 @@ func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, back
 
 	vsMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&vs)
 	if err != nil {
-		return nil, nil, errors.WithStack(err)
+		return nil, nil, "", nil, errors.WithStack(err)
 	}
 
 	p.Log.Infof("Returning from VolumeSnapshotBackupItemAction with %d additionalItems to backup", len(additionalItems))
 	for _, ai := range additionalItems {
 		p.Log.Debugf("%s: %s", ai.GroupResource.String(), ai.Name)
 	}
 
-	return &unstructured.Unstructured{Object: vsMap}, additionalItems, nil
+	// The operationID is of the form <namespace>/<volumesnapshot-name>
+	operationID := vs.Namespace + "/" + vs.Name
+	itemToUpdate := []velero.ResourceIdentifier{
+		{
+			GroupResource: kuberesource.VolumeSnapshots,
+			Namespace:     vs.Namespace,
+			Name:          vs.Name,
+		},
+	}
+
+	return &unstructured.Unstructured{Object: vsMap}, additionalItems, operationID, itemToUpdate, nil
+}
+
+func (p *VolumeSnapshotBackupItemAction) Name() string {
+	return "VolumeSnapshotBackupItemAction"
+}
+
+func (p *VolumeSnapshotBackupItemAction) Progress(operationID string, backup *velerov1api.Backup) (velero.OperationProgress, error) {
+	progress := velero.OperationProgress{}
+	if operationID == "" {
+		return progress, biav2.InvalidOperationIDError(operationID)
+	}
+	// The operationID is of the form <namespace>/<volumesnapshot-name>
+	operationIDParts := strings.Split(operationID, "/")
+	if len(operationIDParts) != 2 {
+		p.Log.Errorf("invalid operation ID %s", operationID)
+		return progress, fmt.Errorf("invalid operation ID %s", operationID)
+	}
+
+	_, snapshotClient, err := util.GetClients()
+	if err != nil {
+		return progress, errors.WithStack(err)
+	}
+
+	vs, err := snapshotClient.SnapshotV1().VolumeSnapshots(operationIDParts[0]).Get(context.Background(), operationIDParts[1], metav1.GetOptions{})
+	if err != nil {
+		p.Log.Errorf("error getting volumesnapshot %s/%s: %s", operationIDParts[0], operationIDParts[1], err.Error())
+		return progress, errors.WithStack(err)
+	}
+
+	if vs.Status == nil {
+		return progress, nil
+	}
+
+	if vs.Status.CreationTime != nil {
+		progress.Started = vs.Status.CreationTime.Time
+	}
+
+	if boolptr.IsSetToTrue(vs.Status.ReadyToUse) {
+		progress.Completed = true
+	} else if vs.Status.Error != nil {
+		progress.Completed = true
+		if vs.Status.Error.Time != nil {
+			progress.Updated = vs.Status.Error.Time.Time
```

Comment:
It might be clearer if progress.Updated is always set to current time rather than to error time (and leaving it blank if no error). Generally, the intent of updated is so that users can know when the last update occurred (and, for completed operations, when the completion field was set, since we won't call Progress again once completed returns as true).

Reply:
Thanks for pointing this out. Removed the Updated time setting.

Reply:
Also, remove the start time, because VolumeSnapshot has no start time for the uploading phase;
VolumeSnapshot creation time is not accurate for the operation. It can be earlier than the operation creation time.

Reply:
If we don't have a real start time, I wonder whether it makes sense to use the current time as the start time (on first progress call)? I just wonder whether an empty start time might break something -- in any case we should make sure that all of the velero core logic doesn't break if start time is empty, since there's no guarantee that a plugin would fill it in.

Reply:
In the CSI case, the `Progress()` function cannot tell whether it is running the first time.

I can only find one place to store the start time. The `operationID` parameter in `Progress()` function.
Setting it to the format `operationID := vs.Namespace + "/" + vs.Name + "/" + startedTime` can make it work.